scripts-dev/mypy_synapse_plugin.py

```diff
@@ -58,7 +58,9 @@ def get_method_signature_hook(
 
 
 def _get_true_return_type(signature: CallableType) -> mypy.types.Type:
-    """Get the "final" return type of a maybe async/Deferred."""
+    """
+    Get the "final" return type of a callable which might return returned an Awaitable/Deferred.
+    """
```

Comment:
Aside: If we have function returning e.g. `Deferred[Deferred[T]]`, it looks like this function doesn't loop and extract the inner T. But we shouldn't be doing that in the first place.

(Would such a type even be legitimate?)

Reply:
> (Would such a type even be legitimate?)

I think it wouldn't make sense? Please don't do that.

Reply:
I'm not sure that's even really doable since a `Deferred` returning a `Deferred` will wait on the inner `Deferred`? https://docs.twistedmatrix.com/en/stable/core/howto/defer.html#chaining-deferreds

Anyway, that interaction always confuses me and I'm willing to not support it.